### PR TITLE
Field `status.hostIPs` added for Pod

### LINKED ISSUE #85443: DualStack: Support pod env source get host IPV6
**What would you like to be added**:
We would like to set host IPV6 address as an env var in pod spec.
**Why is this needed**:
In IPV4 scenario, we use EnvVarSource to get status.hostIP and set it as a env var in pod. Can we support IPV6 as IPV4? Or do we have any other better solution?

### LINKED ISSUE #101565: [Feature request] Field `status.hostIPs` added for Pod
#### What would you like to be added:

- Field `status.hostIPs` added for Pod
- Downward API support for status.hostIPs

#### Why is this needed:

If the Pod runs on an enabled IPv6 dual stack node, hope to get the IPv6 address of node 

``` yaml 
apiVersion: v1
kind: Pod
metadata:
  name: test
spec:
  restartPolicy: Never
  containers:
  - args:
    - |
     env | grep IP
    command:
    - /bin/sh
    - -c
    image: busybox:latest
    imagePullPolicy: IfNotPresent
    name: test
    env:
      - name: HOST_IP
        valueFrom:
          fieldRef: 
            fieldPath: status.hostIP
      - name: POD_IP
        valueFrom:
          fieldRef: 
            fieldPath: status.podIP
      - name: POD_IPS
        valueFrom:
          fieldRef: 
            fieldPath: status.podIPs
```

```
POD_IP=10.88.36.205
HOST_IP=10.7.111.12
POD_IPS=10.88.36.205,2001:4860:4860::ad55
```

#### Want happen:

Add environment value from field `status.hostIPs`, which contains ipv6.
``` yaml
      - name: HOST_IPS
        valueFrom:
          fieldRef: 
            fieldPath: status.hostIPs
```

```
HOST_IPS=10.7.111.12,[want a ipv6 address]
```

https://github.com/kubernetes/kubernetes/blob/50e319767cc4aad5c6ff56a8d002c406cdcc3e0e/pkg/util/node/node.go#L124-L134

### PR DESCRIPTION
**Pull request by wzshiming**
#### What type of PR is this?

/kind feature
/sig node
/sig network

#### What this PR does / why we need it:

https://github.com/kubernetes/enhancements/pull/2661
If the Pod runs on an enabled IPv6 dual stack node, hope to get the IPv6 address of node

#### Which issue(s) this PR fixes:

Original PR #101566
Fixes #101565
Fixes #85443

#### Does this PR introduce a user-facing change?

``` release-note
Alpha support for the field `status.hostIPs` was added to the Pod API, and downward API support for `status.hostIPs` as added. To use either, the alpha feature gate `PodHostIPs` is required to be enabled.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/pull/2661
- [Website PR]: https://github.com/kubernetes/website/pull/42003
```

### CODE REVIEW

File: pkg/apis/core/validation/validation.go [hunk 1]
```diff
@@ -2423,6 +2427,16 @@ func validateObjectFieldSelector(fs *core.ObjectFieldSelector, expressions *sets
...
+func validateDownwardAPIHostIPs(fieldSel *core.ObjectFieldSelector, fldPath *field.Path, opts PodValidationOptions) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if !opts.AllowHostIPsField {
+		if fieldSel.FieldPath == "status.hostIPs" {
+			allErrs = append(allErrs, field.Invalid(fldPath, "status.hostIPs", "not allowed when feature gate 'PodHostIPs' is not enabled"))
+		}
+	}
+	return allErrs
+}
+
```
Comment (wzshiming) [inline pkg/apis/core/validation/validation.go:2439]: https://github.com/kubernetes/kubernetes/pull/101566#discussion_r830937885
https://github.com/kubernetes/kubernetes/pull/101566#discussion_r838719070
https://github.com/kubernetes/kubernetes/pull/101566#discussion_r837686885

Turn on this feature and has a Pod that uses this feature then turn off this feature, here needs to be allowed.

``` yaml
apiVersion: v1
kind: Pod
metadata:
  name: test-pod
spec:
  containers:
    - name: test-container
      image: gcr.io/google_containers/busybox
      command: [ "/bin/sh", "-c", "env" ]
      env:
        - name: MY_POD_HOST_IPS
          valueFrom:
            fieldRef:
              fieldPath: status.hostIPs
  restartPolicy: Never
```
There is already such a pod running, then turning off this feature affects the existing pod.

**Comment by wzshiming** [comment]
Types conversion
https://github.com/kubernetes/kubernetes/pull/101566/commits/8e347718e3b91c9f1cd2bcb83ed70c8c6ea434a3

Add dropDisabledStatusFields
https://github.com/kubernetes/kubernetes/pull/101566/commits/b3a82c6a5ea2c3297b5d84ed215c967be6f0b7b3

Add HostIPs by Kubelet
https://github.com/kubernetes/kubernetes/pull/101566/commits/ac9d7203da982bbf257e2131cbcf07479445185a

The reason why the original load is too large should be. It should be that when Feature is disabled, the Conversion section fills HostIPs and then discards in DropDisableDStatusfields.
Remove the conversion section according to my understanding, it should be avoided.

I think it is enough to fill Hostips by Kublet when enabling features.

**Comment by liggitt** [comment]
is there a specific delta that shows the differences between this PR and the original? are there added tests that would have caught the issues with the previous PR?

File: pkg/apis/core/validation/validation.go [hunk 2]
```diff
@@ -3512,6 +3512,51 @@ func validatePodIPs(pod *core.Pod) field.ErrorList {
 	return allErrs
 }
 
+// validateHostIPs validates IPs in pod status
+func validateHostIPs(pod *core.Pod) field.ErrorList {
```
Comment (aojea) [inline pkg/apis/core/validation/validation.go:3516]: unit test?

File: pkg/apis/core/types.go [hunk 3]
```diff
@@ -3356,9 +3360,21 @@ type PodStatus struct {
 	// give the resources on this node to a higher priority pod that is created after preemption.
 	// +optional
 	NominatedNodeName string
+
+	// HostIP holds the IP address of the host to which the pod is assigned. Empty if the pod has not started yet.
+	// A pod can be assigned to a node that has a problem in kubelet which in turns mean that HostIP will
+	// not be updated even if there is a node is assigned to pod
 	// +optional
 	HostIP string
```
Comment (aojea) [inline pkg/apis/core/types.go:3368]: @thockin @liggitt  this is the part I didn't understand (I can't find the discussion) , when we did pluralize PodIPs the singular was obtained from the plural and removed from the type

Comment (aojea) [reply pkg/apis/core/types.go:3368]: https://github.com/kubernetes/kubernetes/blob/a83cc51a19d1b5f2b2d3fb75574b04f587ec0054/pkg/apis/core/v1/conversion.go#L256-L287

File: pkg/kubelet/kubelet_pods_test.go [hunk 4]
```diff
@@ -2470,6 +2473,10 @@ func TestConvertToAPIContainerStatuses(t *testing.T) {
 }
 
 func Test_generateAPIPodStatus(t *testing.T) {
```
Comment (aojea) [inline pkg/kubelet/kubelet_pods_test.go:2475]: we need some tests exercising mulitple HostIPs

File: pkg/kubelet/kubelet_pods.go [hunk 5]
```diff
@@ -826,6 +826,19 @@ func (kl *Kubelet) podFieldSelectorRuntimeValue(fs *v1.ObjectFieldSelector, pod
 			return "", err
 		}
 		return hostIPs[0].String(), nil
+	case "status.hostIPs":
```
Comment (aojea) [inline pkg/kubelet/kubelet_pods.go:829]: unit test on https://github.com/kubernetes/kubernetes/blob/b1f91bd510b680e7d455623225ca2da5804ad3e2/pkg/kubelet/kubelet_pods_test.go#L293

File: test/e2e_node/dual_stack_host_ips.go [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+	"k8s.io/kubernetes/test/e2e/network/common"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+	netutils "k8s.io/utils/net"
+)
+
+var _ = common.SIGDescribe("Dual Stack Host IP [Feature:PodHostIPs]", func() {
+	f := framework.NewDefaultFramework("dualstack")
+
+	ginkgo.Context("when creating a Pod, it has PodHostIPs feature", func() {
+		tempSetCurrentKubeletConfig(f, func(initialConfig *kubeletconfig.KubeletConfiguration) {
```
Comment (aojea) [inline test/e2e_node/dual_stack_host_ips.go:46]: not familiar with the e2e_node framework, @endocrimes does this enables the feature gate dynamically?

File: test/e2e_node/dual_stack_host_ips.go [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+	e2eskipper "k8s.io/kubernetes/test/e2e/framework/skipper"
+	"k8s.io/kubernetes/test/e2e/network/common"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+	netutils "k8s.io/utils/net"
+)
+
+var _ = common.SIGDescribe("Dual Stack Host IP [Feature:PodHostIPs]", func() {
+	f := framework.NewDefaultFramework("dualstack")
+
+	ginkgo.Context("when creating a Pod, it has PodHostIPs feature", func() {
```
Comment (aojea) [inline test/e2e_node/dual_stack_host_ips.go:45]: are the titles of the tests mixed? this has the feature gate to false

File: test/e2e_node/dual_stack_host_ips.go [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2enode "k8s.io/kubernetes/test/e2e/framework/node"
+	e2eskipper "k8s.io/kubernetes/test/e2e/framework/skipper"
+	"k8s.io/kubernetes/test/e2e/network/common"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+	netutils "k8s.io/utils/net"
+)
+
+var _ = common.SIGDescribe("Dual Stack Host IP [Feature:PodHostIPs]", func() {
+	f := framework.NewDefaultFramework("dualstack")
```
Comment (aojea) [inline test/e2e_node/dual_stack_host_ips.go:43]: these tests are for pods with hostNetwork false, we should add another 2 tests like this but setting hostNetwork: true on the pods

File: test/e2e_node/dual_stack_host_ips.go [hunk 9]
```diff
@@ -137,5 +141,58 @@ var _ = common.SIGDescribe("Dual Stack Host IP [Feature:PodHostIPs]", func() {
 			err = podClient.Delete(context.TODO(), pod.Name, *metav1.NewDeleteOptions(30))
 			framework.ExpectNoError(err, "failed to delete pod")
 		})
+
+		ginkgo.It("should provide hostIPs as an env var", func() {
```
Comment (aojea) [inline test/e2e_node/dual_stack_host_ips.go:145]: do we need to split this commit ? or is just for make the review easier?

File: pkg/apis/core/validation/validation.go [hunk 10]
```diff
@@ -3512,6 +3512,51 @@ func validatePodIPs(pod *core.Pod) field.ErrorList {
...
+	}
+
+	// if we have more than one Pod.PodIP then
+	// - validate for dual stack
+	// - validate for duplication
+	if len(pod.Status.HostIPs) > 1 {
+		hostIPs := make([]string, 0, len(pod.Status.HostIPs))
+		for _, podIP := range pod.Status.HostIPs {
+			hostIPs = append(hostIPs, podIP.IP)
+		}
```
Comment (aojea) [inline pkg/apis/core/validation/validation.go:3535]: s/podIP/hostIP/ in several places

File: test/e2e_node/dual_stack_host_ips.go (hunk 7 above)
Comment (wzshiming) [reply test/e2e_node/dual_stack_host_ips.go:45]: Yes, All tests for a feature even if they're disabled should be here

File: test/e2e_node/dual_stack_host_ips.go (hunk 9 above)
Comment (wzshiming) [reply test/e2e_node/dual_stack_host_ips.go:234]: I merged the two commits for e2e

**Comment by wzshiming** [comment]
> is there a specific delta that shows the differences between this PR and the original? are there added tests that would have caught the issues with the previous PR?

The difference is that the commit https://github.com/kubernetes/kubernetes/commit/8e347718e3b91c9f1cd2bcb83ed70c8c6ea434a3 is removed, and added some tests when the Feature is disabled to ensure that the HostIPs field does not work.

**Comment by pacoxu** [comment]
https://github.com/pacoxu/kubernetes/blob/21238c2593bbda023690258ecf1fec7a29d9dd1a/pkg/kubelet/container/helpers.go#L335-L372

@aojea 
Should kubelet support the hostPort for two HostIPs?

This is not mentioned in the KEP.

**Comment by wzshiming** [comment]
@thockin
Hi
Could you have time to take a look.

**Comment by wzshiming** [comment]
The `go.mod` modified and failed tests are not related to this PR. I'll rebase when the problem is fixed after a while.

**Comment by dims** [comment]
If you still need this PR then please rebase, if not, please close the PR

File: pkg/api/pod/util.go [hunk 11]
```diff
@@ -456,6 +461,57 @@ func GetValidationOptionsFromPodSpecAndMeta(podSpec, oldPodSpec *api.PodSpec, po
...
+		}
+	}
+	return false
+}
+
+func hasUsedDownwardAPIFieldPathWithVolume(volume *api.Volume, fieldPath string) bool {
+	if volume == nil {
+		return false
+	}
+	if volume.DownwardAPI != nil {
```
Comment (thockin) [inline pkg/api/pod/util.go:490]: What about the `projected` volume?

**Review by thockin — COMMENTED** [review]
Code LGTM mostly.  I am not super familiar with node e2e so I think someone from node needs to take a look.

@mrunalp @bobbypage @endocrimes  ?

**Comment by liggitt** [comment]
pull-kubernetes-e2e-gce-cos-alpha-features failing all runs on this PR on mostly network tests when the job is generally green looks suspicious... worth digging into why that is failing

File: staging/src/k8s.io/api/core/v1/types.go [hunk 12]
```diff
@@ -4031,10 +4037,23 @@ type PodStatus struct {
 	// +optional
 	NominatedNodeName string `json:"nominatedNodeName,omitempty" protobuf:"bytes,11,opt,name=nominatedNodeName"`
 
-	// IP address of the host to which the pod is assigned. Empty if not yet scheduled.
+	// hostIP holds the IP address of the host to which the pod is assigned. Empty if the pod has not started yet.
+	// A pod can be assigned to a node that has a problem in kubelet which in turns mean that HostIP will
+	// not be updated even if there is a node is assigned to pod
```
Comment (aojea) [inline staging/src/k8s.io/api/core/v1/types.go:4042]: I think that this only can happen if the node has been deleted or the kublet has lost connection to the apiserver, this comment is a bit confusing

File: pkg/apis/core/validation/validation.go [hunk 13]
```diff
@@ -3738,6 +3738,51 @@ func validatePodIPs(pod *core.Pod) field.ErrorList {
...
+	hostIPsField := field.NewPath("status", "hostIPs")
+
+	// all HostPs must be valid IPs
+	for i, hostIP := range pod.Status.HostIPs {
+		for _, msg := range validation.IsValidIP(hostIP.IP) {
+			allErrs = append(allErrs, field.Invalid(hostIPsField.Index(i), hostIP.IP, msg))
+		}
+	}
+
+	// if we have more than one Pod.PodIP then
```
Comment (aojea) [inline pkg/apis/core/validation/validation.go:3768]: s/Pod.PodIP/Pod.HostIP/

File: pkg/apis/core/validation/validation.go [hunk 14]
```diff
@@ -3738,6 +3738,51 @@ func validatePodIPs(pod *core.Pod) field.ErrorList {
...
+		}
+
+		// We only support one from each IP family (i.e. max two IPs in this list).
+		if !dualStack || len(hostIPs) > 2 {
+			allErrs = append(allErrs, field.Invalid(hostIPsField, pod.Status.HostIPs, "may specify no more than one IP for each IP family"))
+		}
+
+		// There should be no duplicates in list of Pod.HostIPs
+		seen := sets.String{}
+		for i, hostIP := range pod.Status.HostIPs {
```
Comment (aojea) [inline pkg/apis/core/validation/validation.go:3789]: you create a new slice to get the hostIPs and now you parse the pod.Status.HostIPs, do we need the allocation or should we use the allocated slice here?

File: test/e2e_node/dual_stack_host_ips.go [hunk 15]
```diff
@@ -0,0 +1,289 @@
...
+	f := framework.NewDefaultFramework("dualstack")
+
+	ginkgo.Context("when creating a Pod, it has PodHostIPs feature", func() {
+		tempSetCurrentKubeletConfig(f, func(initialConfig *kubeletconfig.KubeletConfiguration) {
+			initialConfig.FeatureGates = map[string]bool{
+				string(kubefeatures.PodHostIPs): false,
+			}
+		})
+		ginkgo.It("should create pod, add host ips is empty", func() {
+			e2eskipper.SkipIfFeatureGateEnabled(kubefeatures.PodHostIPs)
```
Comment (aojea) [inline test/e2e_node/dual_stack_host_ips.go:54]: this doesn't work for e2e , maybe it works for e2e_node, but you have to tag the test as `[Feature:..]`

File: pkg/apis/core/fuzzer/fuzzer.go [hunk 16]
```diff
@@ -91,6 +91,10 @@ var Funcs = func(codecs runtimeserializer.CodecFactory) []interface{} {
 				s.EnableServiceLinks = &enableServiceLinks
 			}
 		},
+		func(s *core.PodStatus, c fuzz.Continue) {
```
[the remark this answers is not in the document]
Comment (wzshiming) [reply pkg/apis/core/fuzzer/fuzzer.go:94]: Here the PodIP have been completely replaced by PodIPs, and even the fields have been removed, so there is no need for this.

File: pkg/api/pod/util.go (hunk 11 above)
Comment (wzshiming) [reply pkg/api/pod/util.go:490]: This part references the other DownwardAPI's `projected`

https://github.com/kubernetes/kubernetes/blob/05ecc3d602f2624016ef34bff753503deae98a66/pkg/api/pod/util.go#L298-L327

File: pkg/apis/core/validation/validation.go (hunk 14 above)
Comment (wzshiming) [reply pkg/apis/core/validation/validation.go:3789]: This is just validation that the HostIPs values are not duplicated.
Sorry did I miss something? I don't quite understand what you mean.

File: test/e2e_node/dual_stack_host_ips.go (hunk 15 above)
Comment (wzshiming) [reply test/e2e_node/dual_stack_host_ips.go]: This part I was previously referring to other e2e, now seems to be no longer necessary. updated.

File: pkg/apis/core/validation/validation.go (hunk 14 above)
Comment (wzshiming) [reply pkg/apis/core/validation/validation.go:3789]: Oh, I read the context carefully, I understand what you mean.

File: staging/src/k8s.io/api/core/v1/types.go (hunk 12 above)
Comment (wzshiming) [reply staging/src/k8s.io/api/core/v1/types.go:4039]: I have no hard opinion on it, this I remember is based on the last comment to change.

File: pkg/apis/core/validation/validation.go [hunk 17]
```diff
@@ -3736,6 +3752,49 @@ func validatePodIPs(pod *core.Pod) field.ErrorList {
 	return allErrs
 }
 
+// validateHostIPs validates IPs in pod status
+func validateHostIPs(pod *core.Pod) field.ErrorList {
```
Comment (thockin) [inline pkg/apis/core/validation/validation.go:3756]: Where do we ensure that `HostIPs[0]` == `HostIP` ?

File: pkg/api/pod/util.go (hunk 11 above)
Comment (thockin) [reply pkg/api/pod/util.go:490]: I don't understand your answer.

Therte's another volume type, "projected" which has a different way to use this informtsaion that isn't being checked here.

**Comment by wzshiming** [comment]
> pull-kubernetes-e2e-gce-cos-alpha-features failing all runs on this PR on mostly network tests when the job is generally green looks suspicious... worth digging into why that is failing

https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/109616/pull-kubernetes-e2e-gce-cos-alpha-features/1607060890820022272/artifacts/e2e-81843c4066-92f64-master/kube-apiserver.log

``` console
E1225 17:47:50.611259       9 fieldmanager.go:210] "[SHOULD NOT HAPPEN] failed to update managedFields" err="failed to convert new object (non-graceful-shutdown-6830/sts-pod-gcepd-d969d9679-5jkzs; /v1, Kind=Pod) to smd typed: .status.hostIPs: element 0: associative list without keys has an element that's a map type" VersionKind="/, Kind=" namespace="non-graceful-shutdown-6830" name="sts-pod-gcepd-d969d9679-5jkzs"
```

I found the reason for the `pull-kubernetes-e2e-gce-cos-alpha-features` test failure, and it looks like it's caused by a failed update here.

**Comment by wzshiming** [comment]
It seems to be caused by `// +listType=set`, I removed it and added `API rule violation` and it works fine.

**Comment by liggitt** [comment]
> It seems to be caused by `// +listType=set`, I removed it and added `API rule violation` and the test passed.

please don't add new violations to the allowlist... listType=atomic would work as well and not require adding that

**Comment by liggitt** [comment]
> E1225 17:47:50.611259       9 fieldmanager.go:210] "[SHOULD NOT HAPPEN] failed to update managedFields" err="failed to convert new object (non-graceful-shutdown-6830/sts-pod-gcepd-d969d9679-5jkzs; /v1, Kind=Pod) to smd typed: .status.hostIPs: element 0: associative list without keys has an element that's a map type" VersionKind="/, Kind=" namespace="non-graceful-shutdown-6830" name="sts-pod-gcepd-d969d9679-5jkzs"

@apelisse - shouldn't a linter have caught this at openapi-time, not at runtime?

**Comment by apelisse** [comment]
Yes, it's exactly the same class of error that we had last month, tracked by https://github.com/kubernetes/kube-openapi/issues/343

**Comment by wzshiming** [comment]
Done

File: pkg/api/pod/util.go [hunk 18]
```diff
@@ -456,6 +461,55 @@ func GetValidationOptionsFromPodSpecAndMeta(podSpec, oldPodSpec *api.PodSpec, po
...
+		}
+	}
+	return false
+}
+
+func hasUsedDownwardAPIFieldPathWithVolume(volume *api.Volume, fieldPath string) bool {
+	if volume == nil || volume.DownwardAPI == nil {
+		return false
+	}
+	for _, file := range volume.DownwardAPI.Items {
```
Comment (thockin) [inline pkg/api/pod/util.go:490]: Comment got lost: 

There's another volume type, "projected" which has a different way to use this information that isn't being checked here.

File: pkg/apis/core/validation/validation.go [hunk 19]
```diff
@@ -3736,6 +3752,58 @@ func validatePodIPs(pod *core.Pod) field.ErrorList {
...
+	allErrs := field.ErrorList{}
+
+	if len(pod.Status.HostIPs) == 0 {
+		return allErrs
+	}
+
+	hostIPsField := field.NewPath("status", "hostIPs")
+
+	// hostIP must be equal to hostIPs[0].IP
+	if pod.Status.HostIP != pod.Status.HostIPs[0].IP {
```
Comment (thockin) [inline pkg/apis/core/validation/validation.go:3766]: I know I flagged this before, but I'm looking for where this is done for podIPs and not finding it.  I need to revisit this.

**Review by thockin — COMMENTED** [review]
sending, but will revisit

File: pkg/apis/core/types.go [hunk 20]
```diff
@@ -3496,9 +3499,21 @@ type PodStatus struct {
...
 
+	// HostIPs holds the IP addresses allocated to the host. If this field is specified, the first entry must
+	// match the hostIP field. This list is empty if the pod has not started yet.
+	// A pod can be assigned to a node that has a problem in kubelet which in turns means that HostIPs will
+	// not be updated even if there is a node is assigned to this pod.
+	// match the hostIP field. This list is empty if no IPs have been allocated yet.
+	// +optional
+	HostIPs []HostIP
```
Comment (thockin) [inline pkg/apis/core/types.go:3515]: @wzshiming Do you recall why we decided to add a field to internal version, rather than do it the same way we did PodIPs?  That puts sopme code into conversion but is easier overall to manage, I think...

Comment (wzshiming) [reply pkg/apis/core/types.go:3515]: As far as I can remember, there was a time when `PodIP` and `PodIPs` coexisted in the internal version, but then `PodIP` was gradually replaced, and finally deleted.

File: pkg/apis/core/validation/validation.go (hunk 19 above)
Comment (wzshiming) [reply pkg/apis/core/validation/validation.go:3766]: In the original patch, I fully referenced `podIP` and `podIPs`, and I changed a lot after multiple rounds of review, but I also remember that `podIPs` didn't have that. But I think there should be no problem to add it in as well.

I don't have a hard opinion.

**Comment by thockin** [comment]
The advantage is that they "feel" similar and should probably be
implemented the same way.

Disagree?

The disadvantage is that it's a little more complicated - attaching to the
conversion logic

File: pkg/apis/core/types.go (hunk 20 above)
Comment (aojea) [reply pkg/apis/core/types.go:3515]: I asked this here https://github.com/kubernetes/kubernetes/pull/109616#discussion_r858374927
I'm almost sure we replaced PodIP directly

Comment (liggitt) [reply pkg/apis/core/types.go:3515]: trying to page in all the history around that change, let's make sure whatever we do applies the lessons learned in:
* https://github.com/kubernetes/kubernetes/pull/73977
* https://github.com/kubernetes/kubernetes/issues/88494
  * https://github.com/kubernetes/kubernetes/pull/88505
* https://github.com/kubernetes/kubernetes/pull/88988
* https://github.com/kubernetes/kubernetes/issues/89888
  * https://github.com/kubernetes/kubernetes/pull/89914
* https://github.com/kubernetes/kubernetes/issues/90625
  * https://github.com/kubernetes/kubernetes/pull/90628

Comment (wzshiming) [reply pkg/apis/core/types.go:3515]: I reviewed these changes and only #88505 was not included, as this `conversion` function was removed from this work outright after learning the lesson from [#101566 commits](https://github.com/kubernetes/kubernetes/pull/101566/commits/8e347718e3b91c9f1cd2bcb83ed70c8c6ea434a3) about overloading `apiserver` due to repeatedly erasing and restoring `hostIPs` due to `conversion` in a cluster with version-skewed `apiserver` and `kubelet`.

Comment (wzshiming) [reply pkg/apis/core/types.go:3515]: Is there anything else I can do?

**Comment by wzshiming** [comment]
@liggitt
Is there anything more I can do?

**Comment by liggitt** [comment]
> @liggitt
> Is there anything more I can do?

I'll defer to thockin on review of this one

**Comment by thockin** [comment]
I am looking, but we are in the last hour of code frereze :(

File: pkg/apis/core/validation/validation.go (hunk 19 above)
Comment (thockin) [reply pkg/apis/core/validation/validation.go:3777]: I went digging through the history and we are not awesome at being consistent.

For service `clusterIPs` we are very careful to try to handle old clients and new clients.  If you set the new (plural) field you must ALSO set the old (singular) field or we will fail validation.  To do that we retained both fields on internal.  If an old client only sets `clusterIP`, we force `clusterIPs` to match.  If a new client setsThere was a reason I can't recall right now why we don't want new clients to only set the plural, but I can't recall the reason now, and it's not in the code.  @khenidak may recall why we decided that it was important for users of the plural to ALSO set the singular.

For pod `hostIPs` we use conversion and squash them together into just the plural field on internal.  If an old client only sets `hostIP`, we set `hostIPs` for them.  If a client only sets `hostIPs`, we set `hostIP` for them.  If a client sets BOTH, we silently discard the `hostIPs`, when it SHOULD be an error.  So retaining both is more correct.

**Comment by wzshiming** [comment]
@thockin @leilajal 
Hi, could you have time to take a look

**Comment by wzshiming** [comment]
@thockin @liggitt
Hi, could you have time to take a look

File: pkg/apis/core/validation/validation.go [hunk 21]
```diff
@@ -2493,6 +2497,16 @@ func validateObjectFieldSelector(fs *core.ObjectFieldSelector, expressions *sets
 	return allErrs
 }
 
+func validateDownwardAPIHostIPs(fieldSel *core.ObjectFieldSelector, fldPath *field.Path, opts PodValidationOptions) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if !opts.AllowHostIPsField {
+		if fieldSel.FieldPath == "status.hostIPs" {
+			allErrs = append(allErrs, field.Invalid(fldPath, "status.hostIPs", "not allowed when feature gate 'PodHostIPs' is not enabled"))
```
Comment (thockin) [inline pkg/apis/core/validation/validation.go:2504]: `field.Forbidden((fldPath, fieldSel.FieldPath, "may not be set when feature gate 'PodHostIPs' is not enabled"))`

File: pkg/apis/core/validation/validation.go [hunk 22]
```diff
@@ -3774,6 +3790,58 @@ func validatePodIPs(pod *core.Pod) field.ErrorList {
...
+
+	if len(pod.Status.HostIPs) == 0 {
+		return allErrs
+	}
+
+	hostIPsField := field.NewPath("status", "hostIPs")
+
+	// hostIP must be equal to hostIPs[0].IP
+	if pod.Status.HostIP != pod.Status.HostIPs[0].IP {
+		allErrs = append(allErrs, field.Invalid(hostIPsField, pod.Status.HostIPs, "hostIPs[0].IP must be equal to hostIP"))
```
Comment (thockin) [inline pkg/apis/core/validation/validation.go:3805]:
```
field.Invalid(hostIPsField.Index(0).Child("ip"), pod.Status.HostIPs[0].ip, "must be equal to `hostIP`"))
```

File: pkg/kubelet/kubelet_pods.go [hunk 23]
```diff
@@ -826,6 +826,19 @@ func (kl *Kubelet) podFieldSelectorRuntimeValue(fs *v1.ObjectFieldSelector, pod
...
+		if !utilfeature.DefaultFeatureGate.Enabled(features.PodHostIPs) {
```
Comment (thockin) [inline pkg/kubelet/kubelet_pods.go:830]: Don't check the gate here - pods that get through the API checks should work

File: pkg/kubelet/kubelet_pods.go [hunk 24]
```diff
@@ -1615,13 +1628,19 @@ func (kl *Kubelet) generateAPIPodStatus(pod *v1.Pod, podStatus *kubecontainer.Po
...
 	})
-	// set HostIP and initialize PodIP/PodIPs for host network pods
+	// set HostIP/HostIPs and initialize PodIP/PodIPs for host network pods
 	if kl.kubeClient != nil {
 		hostIPs, err := kl.getHostIPsAnyWay()
 		if err != nil {
 			klog.V(4).InfoS("Cannot get host IPs", "err", err)
 		} else {
 			s.HostIP = hostIPs[0].String()
+			if utilfeature.DefaultFeatureGate.Enabled(features.PodHostIPs) {
```
Comment (thockin) [inline pkg/kubelet/kubelet_pods.go:1638]: same

File: test/e2e_node/dual_stack_host_ips.go [hunk 25]
```diff
@@ -0,0 +1,286 @@
...
+	e2epod "k8s.io/kubernetes/test/e2e/framework/pod"
+	e2epodoutput "k8s.io/kubernetes/test/e2e/framework/pod/output"
+	"k8s.io/kubernetes/test/e2e/network/common"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+)
+
+var _ = common.SIGDescribe("Dual Stack Host IP [Feature:PodHostIPs]", func() {
+	f := framework.NewDefaultFramework("dualstack")
+
+	ginkgo.Context("when creating a Pod, it has PodHostIPs feature", func() {
```
Comment (thockin) [inline test/e2e_node/dual_stack_host_ips.go:46]: The string says it has the feature, but you set it to false.  Below it is the opposite.  This is "does not have" and the below is "does have"

File: test/e2e_node/dual_stack_host_ips.go
Comment (thockin) [inline test/e2e_node/dual_stack_host_ips.go:1]: I think I'd just call this "pod_host_ips.go"

File: test/e2e_node/dual_stack_host_ips.go [hunk 26]
```diff
@@ -0,0 +1,286 @@
...
+						FieldRef: &v1.ObjectFieldSelector{
+							APIVersion: "v1",
+							FieldPath:  "status.hostIPs",
+						},
+					},
+				},
+			}
+
+			expectations := []string{
+				fmt.Sprintf("HOST_IPS=%v|%v", e2enetwork.RegexIPv4, e2enetwork.RegexIPv6),
```
Comment (thockin) [inline test/e2e_node/dual_stack_host_ips.go:248]: shouldn't we be able to assert the order, at this point?

Comment (wzshiming) [reply test/e2e_node/dual_stack_host_ips.go]: The order is already ensure in the unit test
https://github.com/kubernetes/kubernetes/pull/109616/commits/3ef7c8778fedbc3069fc5ecc341bd5c5bb90239d#diff-1d1207db1b0a769c3db35da7703a0e7302f325a5efde8057a8ac7c13f1caf46dR732

This case was copied from `host_ip` and verified that it actually works
https://github.com/kubernetes/kubernetes/blob/d05b79c836c9db35ef0a60f0175b3a6bbeedfa7e/test/e2e/common/node/downwardapi.go#L86-L110

File: pkg/kubelet/kubelet_pods.go (hunk 24 above)
Comment (wzshiming) [reply pkg/kubelet/kubelet_pods.go:1648]: Here are the fill IPs, if there is no feature gate, the HostIPs will always be set.

File: pkg/kubelet/kubelet_pods.go (hunk 23 above)
Comment (wzshiming) [reply pkg/kubelet/kubelet_pods.go:834]: This deals with a pod that uses this feature and then turns it off, which will lead to this, and since the pod is already using this downward api we can't reject it, and we shouldn't give the value

**Comment by wzshiming** [comment]
Friendly ping @thockin

**Review by thockin — COMMENTED** [review]
Thanks!

/lgtm
/approve

**Comment by wzshiming** [comment]
Hi, @thockin
PTAL again, it just rebases and `./hack/update-codegen.sh`

**Review by thockin — COMMENTED** [review]
Thanks!

/lgtm
/approve

**Comment by wzshiming** [comment]
Thnks
/unhold

**Comment by pacoxu** [comment]
https://perf-dash.k8s.io/#/?jobname=kubemark-500Nodes&metriccategoryname=E2E&metricname=LoadHighThroughputPodStartup&Metric=pod_startup

No regression this time. 🎉
